src/python_testing/TC_CHANNEL_5_9.py

```diff
@@ -0,0 +1,150 @@
+            cluster=Clusters.Channel,
+            attribute=Clusters.Channel.Attributes.ChannelList
+        )
+
+    async def read_current_channel(self, endpoint):
+        return await self.read_single_attribute_check_success(
+            endpoint=endpoint,
+            cluster=Clusters.Channel,
+            attribute=Clusters.Channel.Attributes.CurrentChannel
+        )
+
+    async def send_change_channel_command(self, endpoint, match_str):
+        return await self.default_controller.SendCommand(
+            nodeid=self.dut_node_id,
+            endpoint=endpoint,
+            payload=Clusters.Channel.Commands.ChangeChannel(match=match_str)
+        )
+
+    def extract_channel_identifier(self, channel):
+        """Returns a prioritized match string from the channel struct as per spec."""
+        # Priority: NAME > Identifier > AffiliateCallSign > CallSign > Major.Minor
+        if getattr(channel, "name", None):
+            return channel.name
+        if getattr(channel, "identifier", None):
+            return channel.identifier
+        if getattr(channel, "affiliateCallSign", None):
+            return channel.affiliateCallSign
+        if getattr(channel, "callSign", None):
+            return channel.callSign
+        if hasattr(channel, "majorNumber") and hasattr(channel, "minorNumber"):
+            return f"{channel.majorNumber}.{channel.minorNumber}"
+        return None
+
+    @async_test_body
+    async def test_TC_CHANNEL_5_9(self):
+        endpoint = self.user_params.get("endpoint", 1)
+
+        # Step 0: Commissioning already done by test-runner setup
+        self.step("0")
+
+        # Step 1: Read the ChannelList attribute and validate it's a list
+        if pics("CHANNEL.S.A0000"):
+            self.step("1")
+            channel_list = await self.read_channel_list(endpoint)
+            asserts.assert_is_instance(channel_list, list, "ChannelList must be a list.")
+            logging.info(f"Available channels: {channel_list}")
+            # Filter out channels that don't have any matchable identifiers
+            valid_channels = [ch for ch in channel_list if self.extract_channel_identifier(ch)]
+            asserts.assert_true(len(valid_channels) >= 2, "Need at least 2 valid channels for test.")
+        else:
+            self.skip_step("1")
+
+        # Step 2: Read the CurrentChannel attribute to store current state before switching
+        if pics("CHANNEL.S.A0002"):
+            self.step("2")
+            current_channel = await self.read_current_channel(endpoint)
+            logging.info(f"CurrentChannel: {current_channel}")
+            current_identifier = self.extract_channel_identifier(current_channel) if current_channel else None
+        else:
+            self.skip_step("2")
+
+        # Step 3: Send ChangeChannel command with a different valid Match string than current
+        if pics("CHANNEL.S.C00.Rsp"):
+            self.step("3")
+            selected_channel = next(
```

Comment:
maybe this is clearer:
```suggestion
            selected_channel = valid_channels[0]
            if self.extract_channel_identifier(selected_channel) == current_identifier:
                selected_channel = valid_channels[1]
```

Although this does assume that the channel identifiers are unique in the list. Is that a requirement? Otherwise the assert of "2 valid channels are required" is also insufficient: we need 2 valid channels with distinct identifiers.